_posts/05-06-01-Internationalization-and-Localization.md

```diff
@@ -0,0 +1,384 @@
+---
+title:   Internationalization and Localization
+isChild: true
+anchor:  i18n_l10n
+---
+
+## Internationalization (i18n) and Localization (l10n) {#i18n_l10n_title}
+
+_Disclaimer for newcomers: i18n and l10n are numeronyms, a kind of abbreviation where numbers are used to shorten
+words - in our case, internationalization becomes i18n and localization, l10n._
+
+First of all, we need to define those two similar concepts and other related things:
+
+- **Internationalization** is when you organize your code so it can be adapted to different languages or regions
+without refactors. This is usually done once - preferably, in the beginning of the project, or else you'll probably
+need some huge changes in the source!
+- **Localization** happens when you adapt the interface (mainly) by translating contents, based on the i18n work done
+before. It usually is done every time a new language or region needs support, and is updated when new interface pieces
+are added, as they need to be available in all supported languages.
+- **Pluralization** defines the rules needed between different languages to interoperate strings containing numbers and 
+counters. For instance, in English when you have only one item, it's singular, and anything different from that is 
+called plural; plural in this language is indicated by adding an S after some words, and sometimes changes parts of it.
+In other languages such as Russian or Serbian there are two plural forms plus the singular one - you may even find
+languages with a total of four, five or six forms, such as Slovenian, Irish or Arabic.
+
+## Common ways to implement
+The easiest way to internationalize PHP software is by using array files and using those strings in templates, such as
+`<h1><?=$TRANS['title_about_page']?></h1>`. This is, however, hardly a recommended way for serious projects, as it poses
+some maintenance issues along the road - some might appear in the very beginning, such as pluralization. So, please,
+don't try this if your project will contain more than a couple of pages.
+
+Some frameworks will sport their own i18n packages. Those usually are a more powerful version of the above approach,
+but including features needed for real localization, such as plural forms and string replacement. You're free to use
+those if you feel like, but you might find bothering to edit array source files, having to deal with pure code issues
+(such as string scaping and so on). The main pro here is integration with the environment you're using - the framework
+is called _full-stack_ for a reason, right?
+
+However, the most classic way and often taken as reference for i18n and l10n is a [Unix tool called `gettext`][gettext].
+It dates back to 1995 and is still the most complete implementation for translating software. It is pretty easy to get
```

Comment:
Yeah, generally we don't just have one tool, we put a few decent options in so people can find the right thing. So a user and gettext and maybe a popular alternative would be two good additions to what you have. If you know of any! 

--
Phil Sturgeon
Sent from my iPhone and there's probably typos because I'm probably at the pub.

> On Jul 20, 2016, at 16:29, Igor Santos <notifications@github.com> wrote:
> 
> In _posts/05-06-01-Internationalization-and-Localization.md:
> 
> > +languages with a total of four, five or six forms, such as Slovenian, Irish or Arabic.
> > +
> > +## Common ways to implement
> > +The easiest way to internationalize PHP software is by using array files and using those strings in templates, such as
> > +`<h1><?=$TRANS['title_about_page']?></h1>`. This is, however, hardly a recommended way for serious projects, as it poses
> > +some maintenance issues along the road - some might appear in the very beginning, such as pluralization. So, please,
> > +don't try this if your project will contain more than a couple of pages.
> > +
> > +Some frameworks will sport their own i18n packages. Those usually are a more powerful version of the above approach,
> > +but including features needed for real localization, such as plural forms and string replacement. You're free to use
> > +those if you feel like, but you might find bothering to edit array source files, having to deal with pure code issues
> > +(such as string scaping and so on). The main pro here is integration with the environment you're using - the framework
> > +is called _full-stack_ for a reason, right?
> > +
> > +However, the most classic way and often taken as reference for i18n and l10n is a [Unix tool called `gettext`][gettext].
> > +It dates back to 1995 and is still the most complete implementation for translating software. It is pretty easy to get
> ok, just to make sure I got it right. What's missing is citing some sort of library like this (awesomely giant) Gettext?
> I liked the idea, it's a great addition :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>